proposals/0035-linalg-matrix.md

```diff
@@ -70,47 +89,758 @@ class Matrix {
 
-  void Store(groupshared ElTy Arr[], uint StartIdx, uint Stride, bool ColMajor);
+  template <typename T, uint K>
+    requires ArithmeticScalar<T>
+  std::enable_if_t<Use == MatrixUse::Accumulator && Scope == MatrixScope::Wave,
+                   void>
+  MultiplyAccumulate(const Matrix<T, M, K, MatrixUse::A, Scope> &,
+                     const Matrix<T, K, N, MatrixUse::B, Scope> &);
 
-  void MultiplyAccumulate(const ref Matrix<T, N, M>);
-  void SumAccumulate(const ref Matrix<T, N, M>);
+  template <typename T, unit K>
+    requires ArithmeticScalar<T>
+  std::enable_if_t<Use == MatrixUse::Accumulator && Scope == MatrixScope::Wave,
+                   void>
+  SumAccumulate(const Matrix<T, M, K, MatrixUse::A, Scope>,
+                const Matrix<T, K, N, MatrixUse::B, Scope>);
+
+  // Cooperative Vector outer product accumulate.
+  template <typename T>
+  std::enable_if_t<Use == MatrixUse::Accumulator, void>
+  OuterProductAccumulate(const vector<T, M> &, const vector<T, N> &);
 };
 
 template <typename T, uint M, uint N, uint K>
-Matrix<T, M, N> Multiply(const ref Matrix<T, M, K>, const ref Matrix<T, K, N>);
+Matrix<T, M, N, MatrixUse::A, MatrixScope::Wave>
+Multiply(const Matrix<T, M, K, MatrixUse::A, MatrixScope::Wave>,
+         const Matrix<T, K, N, MatrixUse::B, MatrixScope::Wave>);
+
+// Cooperative Vector Replacement API
+// Cooperative Vector operates on per-thread vectors multiplying against B
+// matrices.
+
+template <typename OutputElTy, typename InputElTy, uint M, uint K,
+          MatrixComponentType MatrixDT, MatrixScope Scope, bool MatrixTranspose>
+vector<OutputElTy, K>
+Multiply(vector<InputElTy, M> InputVector,
+         Matrix<MatrixDT, M, K, MatrixUse::B, Scope> Matrix);
+
+template <typename OutputElTy, typename InputElTy, typename BiasElTy, uint M,
+          uint K, MatrixComponentType MatrixDT, MatrixScope Scope,
+          bool MatrixTranspose>
+vector<OutputElTy, K>
+MultiplyAdd(vector<InputElTy, M> InputVector,
+            Matrix<MatrixDT, M, K, MatrixUse::B, Scope> Matrix,
+            vector<BiasElTy, K> BiasVector);
 
 } // namespace linalg
 } // namespace hlsl
 ```
 
 ## Detailed design
 
+### HLSL API Concepts
+
+The new HLSL API introduces a new `linalg::Matrix` type which represents an
+opaque matrix object, and contains an intangible handle that refers to the
+allocated matrix.
+
+The `linalg::Matrix` template type is parameterized based on the matrix
+component data type, dimensions, use, and scope. These parameters restrict where
+and how a matrix can be used.
+
+#### Matrix Use
+
+The `Use` parameter of an instance of a `linalg::Matrix` denotes which argument
+it can be in matrix-matrix operations or matrix-vector operations.
+
+There are three matrix usages: `A`, `B`, and `Accumulator`.
+* The `A` matrix usage denotes a matrix that can be the first argument to binary
+  or ternary algebraic operations.
+* The `B` matrix usage denotes a matrix that can the second argument to binary

```

Comment:
Its unclear to me (besides dimensions which are already in the type) why A is only allowed as the first parameter and B the second. Assuming the dimensions fit together why can't they be swapped?